// @maxgraph/core/dist/view/layout/datatypes/GraphHierarchyEdge.js
/*
Copyright 2021-present The maxGraph project Contributors
Copyright (c) 2006-2015, JGraph Ltd
Copyright (c) 2006-2015, Gaudenz Alder

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/
import GraphAbstractHierarchyCell from './GraphAbstractHierarchyCell';
import ObjectIdentity from '../../../util/ObjectIdentity';
/**
 * An abstraction of a hierarchical edge for the hierarchy layout
 *
 * @category Layout
 */
class GraphHierarchyEdge extends GraphAbstractHierarchyCell {
    /**
     * Constructs a hierarchy edge
     *
     * @param edges a list of real graph edges this abstraction represents
     */
    constructor(edges) {
        super();
        /**
         * The node this edge is sourced at
         */
        this.source = null;
        /**
         * The node this edge targets
         */
        this.target = null;
        /**
         * Whether or not the direction of this edge has been reversed
         * internally to create a DAG for the hierarchical layout
         */
        this.isReversed = false;
        this.edges = edges;
        this.ids = [];
        for (let i = 0; i < edges.length; i += 1) {
            this.ids.push(ObjectIdentity.get(edges[i]));
        }
    }
    /**
     * Inverts the direction of this internal edge(s)
     */
    invert() {
        const temp = this.source;
        this.source = this.target;
        this.target = temp;
        this.isReversed = !this.isReversed;
    }
    /**
     * Returns the cells this cell connects to on the next layer up
     */
    getNextLayerConnectedCells(layer) {
        if (this.nextLayerConnectedCells == null) {
            this.nextLayerConnectedCells = [];
            for (let i = 0; i < this.temp.length; i += 1) {
                this.nextLayerConnectedCells[i] = [];
                if (i === this.temp.length - 1) {
                    this.nextLayerConnectedCells[i].push(this.source);
                }
                else {
                    this.nextLayerConnectedCells[i].push(this);
                }
            }
        }
        return this.nextLayerConnectedCells[layer - this.minRank - 1];
    }
    /**
     * Returns the cells this cell connects to on the next layer down
     */
    getPreviousLayerConnectedCells(layer) {
        if (this.previousLayerConnectedCells == null) {
            this.previousLayerConnectedCells = [];
            for (let i = 0; i < this.temp.length; i += 1) {
                this.previousLayerConnectedCells[i] = [];
                if (i === 0) {
                    this.previousLayerConnectedCells[i].push(this.target);
                }
                else {
                    this.previousLayerConnectedCells[i].push(this);
                }
            }
        }
        return this.previousLayerConnectedCells[layer - this.minRank - 1];
    }
    /**
     * Returns true.
     */
    isEdge() {
        return true;
    }
    /**
     * Gets the value of temp for the specified layer
     */
    getGeneralPurposeVariable(layer) {
        return this.temp[layer - this.minRank - 1];
    }
    /**
     * Set the value of temp for the specified layer
     */
    setGeneralPurposeVariable(layer, value) {
        this.temp[layer - this.minRank - 1] = value;
    }
    /**
     * Gets the first core edge associated with this wrapper
     */
    getCoreCell() {
        if (this.edges.length > 0) {
            return this.edges[0];
        }
        return null;
    }
}
export default GraphHierarchyEdge;
